In [4]:
import numpy as np
import pandas as pd
import openpyxl
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

In [6]:
df1=pd.read_csv('application_record.csv')
df2=pd.read_csv('credit_record.csv')

In [7]:
df1.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment      -12005          -4542   
1        Civil marriage   Rented apartment      -12005          -4542   
2               Married  House / apartment      -21474          -1134   
3  Single / not married  House / apartment      -19110          -3051   
4  Single / not married  House / apartment      -19110          -3051   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0           1                1           0           0             NaN   
1           1                1           0           0             NaN   
2           1                0           0           0  Security staff   
3           1                0           1           1     Sales staff   
4           1                0           1           1     Sales staff   

   CNT_FAM_MEMBERS  
0              2.0  
1              2.0  
2              2.0  
3              1.0  
4              1.0

In [8]:
df2.head()

ID  MONTHS_BALANCE STATUS
0  5001711               0      X
1  5001711              -1      0
2  5001711              -2      0
3  5001711              -3      0
4  5001712               0      C

In [9]:
df1.columns

Index(['ID', 'CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN',
       'AMT_INCOME_TOTAL', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE',
       'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'DAYS_BIRTH',
       'DAYS_EMPLOYED', 'FLAG_MOBIL', 'FLAG_WORK_PHONE', 'FLAG_PHONE',
       'FLAG_EMAIL', 'OCCUPATION_TYPE', 'CNT_FAM_MEMBERS'],
      dtype='object')

In [10]:
df2.columns

Index(['ID', 'MONTHS_BALANCE', 'STATUS'], dtype='object')

In [11]:
df1['ID'].nunique()

438510

In [12]:
df2['ID'].nunique()

45985

Here we dont have equal number of records in both data sets and we also need to do lot of feature enginnering so lets get on with it

In [26]:
df1_id = list(df1['ID'].unique())
df2_id = list(df2['ID'].unique())


438510

In [29]:
len(list(set(df1_id) & set(df2_id)))

36457

so we can see there are only 36457 records which have records in both application and credit record data sets so we can only work with these data

In [33]:
df1.dtypes

ID                       int64
CODE_GENDER             object
FLAG_OWN_CAR            object
FLAG_OWN_REALTY         object
CNT_CHILDREN             int64
AMT_INCOME_TOTAL       float64
NAME_INCOME_TYPE        object
NAME_EDUCATION_TYPE     object
NAME_FAMILY_STATUS      object
NAME_HOUSING_TYPE       object
DAYS_BIRTH               int64
DAYS_EMPLOYED            int64
FLAG_MOBIL               int64
FLAG_WORK_PHONE          int64
FLAG_PHONE               int64
FLAG_EMAIL               int64
OCCUPATION_TYPE         object
CNT_FAM_MEMBERS        float64
dtype: object

well we need to clean this data set a bit , lets do that and also little bit of eda before merging with the data

In [35]:
for col in df1:
    if not col == 'AMT_INCOME_TOTAL':
        print(col,df1[col].unique())

ID [5008804 5008805 5008806 ... 6841878 6842765 6842885]
CODE_GENDER ['M' 'F']
FLAG_OWN_CAR ['Y' 'N']
FLAG_OWN_REALTY ['Y' 'N']
CNT_CHILDREN [ 0  1  3  2  4  5 14 19  7  9 12  6]
NAME_INCOME_TYPE ['Working' 'Commercial associate' 'Pensioner' 'State servant' 'Student']
NAME_EDUCATION_TYPE ['Higher education' 'Secondary / secondary special' 'Incomplete higher'
 'Lower secondary' 'Academic degree']
NAME_FAMILY_STATUS ['Civil marriage' 'Married' 'Single / not married' 'Separated' 'Widow']
NAME_HOUSING_TYPE ['Rented apartment' 'House / apartment' 'Municipal apartment'
 'With parents' 'Co-op apartment' 'Office apartment']
DAYS_BIRTH [-12005 -21474 -19110 ...  -8169  -9738 -22257]
DAYS_EMPLOYED [-4542 -1134 -3051 ... -8300 -9020 -4884]
FLAG_MOBIL [1]
FLAG_WORK_PHONE [1 0]
FLAG_PHONE [0 1]
FLAG_EMAIL [0 1]
OCCUPATION_TYPE [nan 'Security staff' 'Sales staff' 'Accountants' 'Laborers' 'Managers'
 'Drivers' 'Core staff' 'High skill tech staff' 'Cleaning staff'
 'Private service staff' 'Cooking sta

Here we can see few data can be converted directly in to 1 and 0 using label encoder

In [39]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df1["CODE_GENDER"]=le.fit_transform(df1["CODE_GENDER"])
df1["FLAG_OWN_CAR"]=le.fit_transform(df1["FLAG_OWN_CAR"])
df1["FLAG_OWN_REALTY"]=le.fit_transform(df1["FLAG_OWN_REALTY"])
df1["FLAG_WORK_PHONE"]=le.fit_transform(df1["FLAG_WORK_PHONE"])
df1["FLAG_PHONE"]=le.fit_transform(df1["FLAG_PHONE"])
df1["FLAG_EMAIL"]=le.fit_transform(df1["FLAG_EMAIL"])


In [41]:
for col in df1:
    if not col == 'AMT_INCOME_TOTAL':
        print(col,df1[col].unique())

ID [5008804 5008805 5008806 ... 6841878 6842765 6842885]
CODE_GENDER [1 0]
FLAG_OWN_CAR [1 0]
FLAG_OWN_REALTY [1 0]
CNT_CHILDREN [ 0  1  3  2  4  5 14 19  7  9 12  6]
NAME_INCOME_TYPE ['Working' 'Commercial associate' 'Pensioner' 'State servant' 'Student']
NAME_EDUCATION_TYPE ['Higher education' 'Secondary / secondary special' 'Incomplete higher'
 'Lower secondary' 'Academic degree']
NAME_FAMILY_STATUS ['Civil marriage' 'Married' 'Single / not married' 'Separated' 'Widow']
NAME_HOUSING_TYPE ['Rented apartment' 'House / apartment' 'Municipal apartment'
 'With parents' 'Co-op apartment' 'Office apartment']
DAYS_BIRTH [-12005 -21474 -19110 ...  -8169  -9738 -22257]
DAYS_EMPLOYED [-4542 -1134 -3051 ... -8300 -9020 -4884]
FLAG_MOBIL [1]
FLAG_WORK_PHONE [1 0]
FLAG_PHONE [0 1]
FLAG_EMAIL [0 1]
OCCUPATION_TYPE [nan 'Security staff' 'Sales staff' 'Accountants' 'Laborers' 'Managers'
 'Drivers' 'Core staff' 'High skill tech staff' 'Cleaning staff'
 'Private service staff' 'Cooking staff' 'Low-ski

In [45]:
df1 =  pd.get_dummies(df1, columns= ["NAME_INCOME_TYPE", "NAME_EDUCATION_TYPE", "NAME_FAMILY_STATUS","NAME_HOUSING_TYPE"])

In [47]:
convert_to = {'C' : 'Good_Debt', 'X' : 'Good_Debt', '0' : 'Good_Debt', '1' : 'Neutral_Debt', '2' : 'Neutral_Debt', '3' : 'Bad_Debt', '4' : 'Bad_Debt', '5' : 'Bad_Debt'}
df2.replace({'STATUS' : convert_to}, inplace=True)

In [48]:
df2

ID  MONTHS_BALANCE     STATUS
0        5001711               0  Good_Debt
1        5001711              -1  Good_Debt
2        5001711              -2  Good_Debt
3        5001711              -3  Good_Debt
4        5001712               0  Good_Debt
...          ...             ...        ...
1048570  5150487             -25  Good_Debt
1048571  5150487             -26  Good_Debt
1048572  5150487             -27  Good_Debt
1048573  5150487             -28  Good_Debt
1048574  5150487             -29  Good_Debt

[1048575 rows x 3 columns]

In [49]:
df2 = df2.value_counts(subset=['ID', 'STATUS']).unstack(fill_value=0)

In [50]:
df2.loc[(df2['Good_Debt'] > df2['Neutral_Debt']), 'CREDIT_APPROVAL_STATUS'] = 1
df2.loc[(df2['Good_Debt'] > df2['Bad_Debt']), 'CREDIT_APPROVAL_STATUS'] = 1
df2.loc[(df2['Neutral_Debt'] > df2['Good_Debt']), 'CREDIT_APPROVAL_STATUS'] = 0
df2.loc[(df2['Neutral_Debt'] > df2['Bad_Debt']), 'CREDIT_APPROVAL_STATUS'] = 1
df2.loc[(df2['Bad_Debt'] > df2['Good_Debt']), 'CREDIT_APPROVAL_STATUS'] = 0
df2.loc[(df2['Bad_Debt'] > df2['Neutral_Debt']), 'CREDIT_APPROVAL_STATUS'] = 0

In [51]:

df2["CREDIT_APPROVAL_STATUS"].value_counts()

1.0    45796
0.0      189
Name: CREDIT_APPROVAL_STATUS, dtype: int64

In [52]:
df2

STATUS   Bad_Debt  Good_Debt  Neutral_Debt  CREDIT_APPROVAL_STATUS
ID                                                                
5001711         0          4             0                     1.0
5001712         0         19             0                     1.0
5001713         0         22             0                     1.0
5001714         0         15             0                     1.0
5001715         0         60             0                     1.0
...           ...        ...           ...                     ...
5150482         0         18             0                     1.0
5150483         0         18             0                     1.0
5150484         0         13             0                     1.0
5150485         0          2             0                     1.0
5150487         0         30             0                     1.0

[45985 rows x 4 columns]

In [53]:
df2['CREDIT_APPROVAL_STATUS'] = df2['CREDIT_APPROVAL_STATUS'].astype('int')